amplipi/app.py

```diff
@@ -897,65 +898,83 @@ def get_ip_info(iface: str = 'eth0') -> Tuple[Optional[str], Optional[str]]:
     return None, None
 
 
-def advertise_service(port, que: Queue):
+def advertise_service(port, event: SyncEvent):
   """ Advertise the AmpliPi api via zeroconf, can be verified with 'avahi-browse -ar'
-      Expected to be run as a seperate process, eg:
+      Expected to be run as a separate process, eg:
 
-          q = Queue()
-          ad = Process(target=amplipi.app.advertise_service, args=(5000, q))
+          event = multiprocessing.Event()
+          ad = Process(target=amplipi.app.advertise_service, args=(5000, event))
           ad.start()
           ...
-          q.put('done')
+          event.set()
           ad.join()
+      NOTE: multiprocessing.Event() is a function that returns a multiprocessing.synchronize.Event type
+      Here the type is aliased to SyncEvent
   """
+  def ok():
+    """ Was a stop requested by the parent process? """
+    return not event.is_set()
+
+  while ok():
+    try:
+      _advertise_service(port, ok)
+    except Exception as exc:
+      if 'change registration' not in str(exc):
+        print(f'Failed to advertise AmpliPi service, retrying in 30s: {exc}')
+        # delay for a bit after a failure
+        delay_ct = 300
+        while ok() and delay_ct > 0:
+          sleep(0.1)
+          delay_ct -= 1
+
+
+def _find_best_iface(ifaces: List[str], ok: Callable) -> Union[Tuple[str, str, str], Tuple[None, None, None]]:
+  # Try to find the best interface to advertise on,
+  # retrying in case DHCP resolution is delayed
+  ip_addr, mac_addr = None, None
+  retry_count = 5
+  while ok() and not (ip_addr and mac_addr) and retry_count > 0:
+    for iface in ifaces:
+      ip_addr, mac_addr = get_ip_info(iface)
+      if ip_addr and mac_addr:
+        return ip_addr, mac_addr, iface
+    sleep(2)  # wait a bit in case this was started before DHCP was started
+    retry_count -= 1
+  return None, None, None
+
+
+def _advertise_service(port: int, ok: Callable) -> None:
+  """ Underlying advertisement, can throw Exceptions when network is not connected """
+
   hostname = f'{gethostname()}.local'
   url = f'http://{hostname}'
 
-  # search for the best interface to advertise on
+  # enumerate interface to advertise on, starting with the pi defaults first
+  # ordering is important, as the pi defaults will be tried first
   ifaces = ['eth0', 'wlan0']  # default pi interfaces
```

Comment:
Do we need to list these? If the issue here is them not showing up right away at boot, would it be better for _find_best_iface to dynamically enumerate the interfaces just like it is doing for IPs?

Reply:
I dumped this in the latest version and switched to depending on the default gateway.